kore/src/Kore/Internal/Predicate.hs

```diff
@@ -119,438 +184,642 @@ import Pretty
     => Sort
+    -- ^ Sort of resulting pattern
     -> Predicate variable
-    -> Predicate variable
-coerceSort sort = fmap (TermLike.fullyOverrideSort sort)
-
-predicateSort :: Predicate variable -> Sort
-predicateSort = termLikeSort . unwrapPredicate
+    -> TermLike variable
+fromPredicate sort = Recursive.fold worker
+  where
+    worker (pat :< predF) = TermLike.setSimplified
+        (Attribute.simplifiedAttribute pat)
+        $ case predF of
+            AndF     (And () t1 t2)       -> TermLike.mkAnd t1 t2
+            BottomF  _                    -> TermLike.mkBottom sort
+            CeilF    (Ceil () () t)       -> TermLike.mkCeil sort t
+            EqualsF  (Equals () () t1 t2) -> TermLike.mkEquals sort t1 t2
+            ExistsF  (Exists () v t)      -> TermLike.mkExists v t
+            FloorF   (Floor () () t)      -> TermLike.mkFloor sort t
+            ForallF  (Forall () v t)      -> TermLike.mkForall v t
+            IffF     (Iff () t1 t2)       -> TermLike.mkIff t1 t2
+            ImpliesF (Implies () t1 t2)   -> TermLike.mkImplies t1 t2
+            InF      (In () () t1 t2)     -> TermLike.mkIn sort t1 t2
+            NotF     (Not () t)           -> TermLike.mkNot t
+            OrF      (Or () t1 t2)        -> TermLike.mkOr t1 t2
+            TopF     _                    -> TermLike.mkTop sort
+
+fromPredicate_
+    :: InternalVariable variable
+    => Predicate variable
+    -> TermLike variable
+fromPredicate_ = fromPredicate predicateSort
 
-{-|'PredicateFalse' is a pattern for matching 'bottom' predicates.
--}
-pattern PredicateFalse :: Predicate variable
+data HasChanged = Changed | NotChanged
+    deriving (Show, Eq)
 
-{-|'PredicateTrue' is a pattern for matching 'top' predicates.
--}
-pattern PredicateTrue :: Predicate variable
+instance Semigroup HasChanged where
+    NotChanged <> x = x
+    Changed <> _ = Changed
 
-pattern PredicateFalse <- GenericPredicate (Recursive.project -> _ :< BottomF _)
-pattern PredicateTrue  <- GenericPredicate (Recursive.project -> _ :< TopF _)
+instance Monoid HasChanged where
+    mempty = NotChanged
 
-pattern PredicateAnd
-    :: Predicate variable -> Predicate variable -> Predicate variable
-pattern PredicateAnd p1 p2 <-
-    GenericPredicate (And_ _ (GenericPredicate -> p1) (GenericPredicate -> p2))
 {-|'isFalse' checks whether a predicate is obviously bottom.
 -}
-isFalse :: TopBottom patt => GenericPredicate patt -> Bool
+isFalse :: TopBottom patt => patt -> Bool
 isFalse = isBottom
 
-binaryOperator
-    :: (HasCallStack, InternalVariable variable)
-    => (TermLike variable -> TermLike  variable -> TermLike  variable)
-    -> Predicate variable -> Predicate variable -> Predicate variable
-binaryOperator operator (GenericPredicate first) (GenericPredicate second) =
-    GenericPredicate (operator first sortedSecond)
-  where
-    sortedSecond = TermLike.fullyOverrideSort (termLikeSort first) second
+isTrue :: TopBottom patt => patt -> Bool
+isTrue = not . isBottom
```

Comment:
I think you want `isTop` here, or else the name `isTrue` is very misleading.

```suggestion
isTrue = isTop
```

Although every predicate is eventually evaluated to `\top` or `\bottom`, not being `\bottom` is not the same as being _known_ `\top`.

Actually, after looking at the places this is used below, I see that you do mean `isFalse = isBottom` and `isTrue = isTop`. In that case, please remove `isFalse` and `isTrue`. Use `isBottom` and `isTop`, which are terms from the Ubiquitous Language.

Reply:
This `isTrue` is never used anywhere as far as I can tell... And the `isFalse` is only used in unit tests (where I now replaced it by `isBottom`).
I now removed them both. I agree that `isTop` shouldn't be defined as such, I don't remember what reasoning I had back when I wrote it, I think it was very early on in the project when I was quite unfamiliar with the code.